src/librustc_trans/back/link.rs

```diff
@@ -980,6 +981,11 @@ fn link_args(cmd: &mut Linker,
         cmd.build_dylib(out_filename);
     }
 
+    if crate_type == config::CrateTypeExecutable &&
+       sess.crt_static() && !t.options.crt_static_allows_dylibs {
```

Comment:
Perhaps we could add a clause here as well for a custom target spec option that's specific to musl? I'd be wary of "breaking" the glibc target by accident for now, and we currently only need `-static` for musl to work, right?